pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,417 @@
+	// only perform work for watched gvks
+	if gvk := instance.GroupVersionKind(); !c.watchedSet.Contains(gvk) {
+		return nil
+	}
+
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces: %w", err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	_, err = c.opa.AddData(ctx, instance)
+	if err != nil {
+		c.syncMetricsCache.AddObject(
+			syncKey,
+			syncutil.Tags{
+				Kind:   instance.GetKind(),
+				Status: metrics.ErrorStatus,
+			},
+		)
+
+		return err
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	// only perform work for watched gvks
+	if gvk := instance.GroupVersionKind(); !c.watchedSet.Contains(gvk) {
+		return nil
+	}
+
+	if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) listAndSyncDataForGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	stopChan := make(chan bool, 1)
+	gvkErrdChan := make(chan schema.GroupVersionKind, 1024)
+	gvksFailingTolist := watch.NewSet()
+
+	gvksFailingToListReconciler := func(stopChan <-chan bool) {
+		for {
+			select {
+			case <-stopChan:
+				return
+			case gvk := <-gvkErrdChan:
```

Comment:
Why do we need to send GVKs over a channel?  Couldn't we just write to a shared structure that's guarded by the mutex?